netkan/netkan/mirrorer.py

```diff
@@ -0,0 +1,294 @@
+import boto3
+import internetarchive
+import tempfile
+import urllib.request
+import requests
+import hashlib
+import logging
+from pathlib   import Path
+from .metadata import Ckan
+
+class Mirrorer:
+
+    CACHE_PATH = Path('/home/netkan/ckan_cache/')
+
+    def __init__(self, queue, timeout, ckan_meta_repo, ia_access, ia_secret, ia_collection):
+        sqs = boto3.resource('sqs')
+        self.sqs_client     = boto3.client('sqs')
+        self.queue          = sqs.get_queue_by_name(QueueName=queue)
+        self.timeout        = timeout
+        self.ckan_meta_repo = ckan_meta_repo
+        self.ia_collection  = ia_collection
+        self.ia_session     = internetarchive.get_session(config = {
+            's3': {
+                'access': ia_access,
+                'secret': ia_secret,
+            }
+        })
+
+    def run(self):
+        while True:
+            messages = self.queue.receive_messages(
+                MaxNumberOfMessages   = 10,
+                MessageAttributeNames = ['All'],
+                VisibilityTimeout     = self.timeout,
+            )
+            if messages:
+                # Check if archive.org is overloaded
+                if self.ia_overloaded():
+                    logging.info('The Internet Archive is overloaded, try again later')
+                    continue
+                # Make sure we have the up to date copy of the metadata for the files we're mirroring
+                logging.info('Updating repo')
+                self.ckan_meta_repo.remotes.origin.pull('master', strategy_option='ours')
+                # Start processing the messages
+                to_delete = []
+                for msg in messages:
+                    path = Path(self.ckan_meta_repo.working_dir, msg.body)
+                    if self.try_mirror(CkanMirror(path)):
+                        # Successfully handled -> OK to delete
+                        to_delete.append(deletion_msg(msg))
+                self.queue.delete_messages(Entries = to_delete)
+
```

Comment:
You'll notice in the indexer we're only building an Indexer object every time we have messages to process, combined with a context manager. There was some method behind the madness, which was to avoid issues with file handles + memory within gitpython.

Reply:
Yeah, that was something I wanted to avoid because the core loop of the indexer is in `cli.py` instead of `indexer.py`. Does it actually solve something? Because in both cases one git repo object is initialized once at startup and then used and re-used as needed during processing. The indexer's core loop does the same thing as this new code: receive a batch of messages, process them, and delete them. I'm not seeing a functional run-time difference.

Reply:
Yeah, I am conscious a lot is happening in the `cli.py`. I think we can refactor a lot of it by creating some click decorators and condensing all the common options. But didn't want to get bogged down with that initially.

https://gitpython.readthedocs.io/en/stable/intro.html#limitations

Reply:
OK, I see now that `indexer.py` does do some extra cleanup in the form of `self.repo.close()`. Will add that here as well, since it appears to be safe (if undocumented):

https://github.com/gitpython-developers/GitPython/blob/6fd090293792884f5a0d05f69109da1c970c3cab/git/repo/base.py#L226-L239